app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.test.ts

```diff
@@ -0,0 +1,756 @@
+  const grantPermissions = jest.fn().mockReturnValue(
+    Object.freeze({
+      [Caip25EndowmentPermissionName]: {
+        id: 'new',
+        parentCapability: Caip25EndowmentPermissionName,
+        caveats: [
+          {
+            type: Caip25CaveatType,
+            value: {
+              requiredScopes: {},
+              optionalScopes: {},
+            },
+          },
+        ],
+      },
+    }),
+  );
+  const requestPermissionApprovalForOrigin = jest.fn().mockResolvedValue({
+    approvedChainIds: ['0x1', '0x5'],
+    approvedAccounts: ['0xdeadbeef'],
+  });
+  const getAccounts = jest.fn().mockReturnValue([]);
+  const response: PendingJsonRpcResponse<Json> = {
+    jsonrpc: '2.0' as const,
+    id: 0,
+  };
+  const handler = (request: unknown) =>
+    requestPermissionsHandler.implementation(
+      request as JsonRpcRequest<[RequestedPermissions]> & { origin: string },
+      response,
+      next,
+      end,
+      {
+        requestPermissionsForOrigin,
+        getPermissionsForOrigin,
+        updateCaveat,
+        grantPermissions,
+        requestPermissionApprovalForOrigin,
+        getAccounts,
+      },
+    );
+
+  return {
+    response,
+    next,
+    end,
+    requestPermissionsForOrigin,
+    getPermissionsForOrigin,
+    updateCaveat,
+    grantPermissions,
+    requestPermissionApprovalForOrigin,
+    getAccounts,
+    handler,
+  };
+};
+
+describe('requestPermissionsHandler', () => {
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  beforeEach(() => {
+    MockMultichain.setEthAccounts.mockImplementation(
+      (caveatValue) => caveatValue,
+    );
+    MockMultichain.setPermittedEthChainIds.mockImplementation(
+      (caveatValue) => caveatValue,
+    );
+  });
+
+  it('returns an error if params is malformed', async () => {
+    const { handler, end } = createMockedHandler();
+
+    const malformedRequest = {
+      ...getBaseRequest(),
+      params: [],
+    };
+    await handler(malformedRequest);
+    expect(end).toHaveBeenCalledWith(
+      invalidParams({ data: { request: malformedRequest } }),
+    );
+  });
+
+  it('requests approval from the ApprovalController for eth_accounts and permittedChains when only eth_accounts is specified in params and origin is not snapId', async () => {
+    const { handler, requestPermissionApprovalForOrigin } =
+      createMockedHandler();
+
+    await handler({
+      ...getBaseRequest(),
+      params: [
+        {
+          [RestrictedMethods.eth_accounts]: {
+            foo: 'bar',
+          },
+        },
+      ],
+    });
+
+    expect(requestPermissionApprovalForOrigin).toHaveBeenCalledWith({
+      [RestrictedMethods.eth_accounts]: {
+        foo: 'bar',
+      },
+      [PermissionNames.permittedChains]: {},
+    });
+  });
+
+  it('requests approval from the ApprovalController for eth_accounts and permittedChains when only permittedChains is specified in params and origin is not snapId', async () => {
```

Comment:
Similar as previous comment, thoughts on making the origin explicit?

Reply:
https://github.com/MetaMask/metamask-extension/pull/27847/commits/f7a09fda54e6e4998b098b2ea84e8557b0e13d3d